v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/PythonExternalTextTransformer.java

```diff
@@ -249,6 +252,46 @@ originalMsg, new String(originalMsg.getBytes(), StandardCharsets.UTF_8))
     }
   }
 
+  public static class RowToStringElementFn extends DoFn<Row, String> {
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      Row row = context.element();
+      String errorMessage = row.getValue("error_message");
+      String stackTrace = row.getValue("stack_trace");
+      if (!Strings.isNullOrEmpty(errorMessage) || !Strings.isNullOrEmpty(stackTrace)) {
+        Object originalElement = row.getValue("original");
+        throw new RuntimeException(
+            String.format(
+                "Error applying UDF to the source record [%s], error message: [%s], stack trace: [%s]",
+                originalElement, errorMessage, stackTrace));
+      }
+
+      Row transformedRow = row.getValue("transformed");
+      context.output(transformedRow.getValue("message"));
+    }
+  }
+
+  public static class RowToTableRowElementFn extends DoFn<Row, TableRow> {
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      Row row = context.element();
+      String errorMessage = row.getValue("error_message");
+      String stackTrace = row.getValue("stack_trace");
+      if (!Strings.isNullOrEmpty(errorMessage) || !Strings.isNullOrEmpty(stackTrace)) {
+        Object originalElement = row.getValue("original");
+        throw new RuntimeException(
```

Comment:
In the UDF we're handling both failsafe and non-failsafe elements. Some elements require that we write out the error to the sink, some don't. Because of this the UDF is failsafe agnostic and returns out as if all of them are failsafe. Then we determine at this point whether or not we want to throw the error. We can mitigate this by breaking out the UDF into multiple different external python UDFs, but the effort for this would be relatively high.